compiler/qsc/src/interpret/stateful.rs

```diff
@@ -406,22 +423,57 @@ impl Interpreter {
 
     /// Runs the given entry expression on a new instance of the environment and simulator,
     /// but using the current compilation.
-    /// # Errors
-    /// If the parsing of the expr fails, an error is returned.
-    /// If the compilation of the expr fails, an error is returned.
-    /// If there is a runtime error when generating code for the expr, an error is returned.
-    /// # Panics
-    /// If internal compiler state is inconsistent, a panic may occur.
     pub fn run(
         &mut self,
         receiver: &mut impl Receiver,
         expr: &str,
         shots: u32,
-    ) -> Result<Vec<LineResult>, Vec<Error>> {
+    ) -> Result<Vec<InterpretResult>, Vec<Error>> {
+        self.run_with_sim(&mut SparseSim::new(), receiver, expr, shots)
+    }
+
+    /// Performs QIR codegen using the given entry expression on a new instance of the environment
+    /// and simulator but using the current compilation.
+    pub fn qirgen(&mut self, expr: &str) -> Result<String, Vec<Error>> {
+        if self.target != TargetProfile::Base {
+            return Err(vec![Error(ErrorKind::TargetMismatch)]);
+        }
+
+        let mut sim = BaseProfSim::new();
+        let mut stdout = std::io::sink();
+        let mut out = GenericReceiver::new(&mut stdout);
+
+        let val = self
+            .run_with_sim(&mut sim, &mut out, expr, 1)?
+            .into_iter()
+            .next()
+            .expect("execution should have exactly one result")?;
```

Comment:
Nit: This is really checking for "at least" one result, not "exactly". Not sure if it's worth adding the extra validation, but the expect message could be updated.

Reply:
Whoops, yes! I had fixed this comment at one point, but then needed to revert some changes and lost this. One option I explored that got lost in the shuffle is using `.last()` instead, since technically I only want the output from the last statement. However, the difference is negligible because we know `run_with_sim` uses the expression parser, so there should only ever be one result from an expression. Still, I think updating the wording and using `.last()` do make the intent clearer.

Reply:
Updated!